src/app/helpers/formTemplate/formTemplate.ts

```diff
@@ -0,0 +1,118 @@
+import { OnInit } from "@angular/core";
+import { ActivatedRoute, Router } from "@angular/router";
+import { FormlyFieldConfig } from "@ngx-formly/core";
+import { ToastrService } from "ngx-toastr";
+import { Observable } from "rxjs";
+import { takeUntil } from "rxjs/operators";
+import { WithFormCheck } from "src/app/guards/form/form.guard";
+import { AbstractModel } from "src/app/models/AbstractModel";
+import { ApiErrorDetails } from "src/app/services/baw-api/api.interceptor.service";
+import { ResolvedModel } from "src/app/services/baw-api/resolver-common";
+import { PageComponent } from "../page/pageComponent";
+
+export abstract class FormTemplate<M extends AbstractModel>
+  extends WithFormCheck(PageComponent)
+  implements OnInit {
+  /**
+   * Form Submission Function Loading
+   */
+  public loading: boolean;
+  /**
+   * Initial setup failed
+   */
+  public failure: boolean;
+  /**
+   * Model to edit using form
+   */
+  public model: M;
+  /**
+   * Extra models stored in data
+   */
+  public models: { [key: string]: AbstractModel } = {};
+  /**
+   * Formly fields
+   */
+  public fields: FormlyFieldConfig[];
+  /**
+   * Success Message
+   */
+  private successMessage: string;
+
+  /**
+   * Customize form template
+   * @param notifications Notifications service
+   * @param route Activated route service
+   * @param router Router service
+   * @param modelKeys Router data model keys (first key is default model)
+   * @param successMsg Success message
+   * @param errorMsg Error message
+   */
+  constructor(
+    protected notifications: ToastrService,
+    protected route: ActivatedRoute,
+    protected router: Router,
+    private modelKeys: string[],
```

Comment:
so this relates to my previous string focused comment. 

I wonder, can we attach to models (maybe to AbsrtactModel) a method that returns the name of the model?
Then all we'd need to pass here is any model or model constructor ...

Reply:
So i already thought about that, and the `kind` key will tell us what type of `AbstractModel` we are dealing with. However i couldnt figure out the next step of typing the variable.